design/how-does-data-distribution-work.md

```diff
@@ -0,0 +1,83 @@
+
+If a new key range is assigned to a storage server, the storage server will receive a private mutation that changes the *serverKeys *(\xff/serverKeys/) and *keyServers* (\xff/keyServers/). Then the server will create transactions, just as an FDB client, to read key-value pairs in the assigned key range and write the data into its local storage engine. 
+
+If a key range is removed from a storage server, similarly the storage server will receive a private mutation that changes the *serverKeys* and *keyServers*. Once the private mutation is processed by the SS, the SS removes data in its versioned data. 
+
+
+## Mechanisms
+
+### How is data distribution initialized?
+
+When a data distribution role is created, it recovers the states of the previous DD from the system keyspace. First, it sets itself as the owner of the moveKeysLock. Then it collects the information of servers and shards, the map between servers and shards, and the replication configuration by reading DD-related system keyspace (i.e., *serverKeys* sub-space). Based on the information, the new DD recreates its components (e.g., servers, teams, and trackers) that matches the states of the previous DD. Trackers will evaluate the healthiness of servers and teams based on the replication policy. Unhealthy servers and teams will be removed and new servers and teams will be created if the replication policy is changed.
+
+### When to move keys?
+
+Keys can be moved from a server to another for several reasons: (a) DD moves keys from overutilized servers to underutilized servers, where a server’s utilization is defined as the server’s disk usage; (b) DD splits or merges shards in order to rebalance the disk usage of servers; (c) DD removes redundant teams when the team number is larger than the desired number; (d) DD repairs the replication factor by duplicate shards from a server to another when servers in a team fail.
+
+Actors are created to monitor the reasons of key movement: (a) MountainChopper and ValleyFiller actors periodically measure a random server team’s utilization and rebalance the server’s keys among other servers; (b) shardMerger and shardSplitter actors take a shard as input and respectively evaluates if the input shard can be merged with its neighboring shards without creating a too big shard and if the shard should be split. Once new shards are created, the actors create the shard’s tracker and send RelocateShard requests to DD’s queue; (c) serverTeamRemover and machineTeamRemover actors periodically evaluate if the number of server teams and machine teams is larger than the desired number. If so, they respectively pick a server team or a machine team to remove based on predefined criteria; (d) teamTracker actor monitors a team’s healthiness. When a server in the team becomes unhealthy, it issues the RelocateShard request to repair the replication factor. The less servers a team has, the higher priority the RelocateShard request will be.
```

Comment:
```suggestion
Actors are created to monitor the reasons of key movement: (a) MountainChopper and ValleyFiller actors periodically measure a random server team’s utilization and rebalances the server’s keys among other servers; (b) shardMerger and shardSplitter actors take a shard as input and respectively evaluates if the input shard can be merged with its neighboring shards without creating a shard too large and if the shard should be split. Once new shards are created, the actors create the shard’s tracker and send RelocateShard requests to DD’s queue; (c) serverTeamRemover and machineTeamRemover actors periodically evaluate if the number of server teams and machine teams is larger than the desired number. If so, they respectively pick a server team or a machine team to remove based on predefined criteria; (d) teamTracker actor monitors a team’s healthiness. When a server in the team becomes unhealthy, it issues the RelocateShard request to repair the replication factor. The less servers a team has, the higher priority the RelocateShard request will be.
```

Reply:
> If so, they respectively pick a server team or a machine team to remove based on predefined criteria;

Perhaps give one or two examples of what criteria is considered. Also consider using a new line for each bullet